packages/plugin-data/src/api/data/soql/reporters.ts

```diff
@@ -0,0 +1,326 @@
+import * as chalk from 'chalk';
+import { get, isString } from '@salesforce/ts-types';
+import { Field, FunctionField, SubqueryField } from './queryFields';
+import { DataSoqlQueryResult } from './dataSoqlQueryTypes';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/plugin-data', 'soql.query');
+
+export class Reporter {
+  protected ux: UX;
+  protected logger: Logger;
+
+  public constructor(ux: UX, logger: Logger) {
+    this.ux = ux;
+    this.logger = logger.child('reporter');
+  }
+
+  public log(...args: string[]): void {
+    this.ux.log(...args);
+  }
+}
+
+export class QueryReporter extends Reporter {
+  protected columns: Field[] = [];
+  protected data: DataSoqlQueryResult;
+
+  public constructor(data: DataSoqlQueryResult, columns: Field[], ux: UX, logger: Logger) {
+    super(ux, logger);
+    this.columns = columns;
+    this.data = data;
+  }
+}
+
+type ParsedFields = {
+  attributeNames: string[];
+  children: string[];
+  aggregates: FunctionField[];
+};
+
+type ColumnAttributes = {
+  key: string;
+  label: string;
+};
+
+export class HumanReporter extends QueryReporter {
+  public constructor(data: DataSoqlQueryResult, columns: Field[], ux: UX, logger: Logger) {
+    super(data, columns, ux, logger);
+  }
+
+  public display(): void {
+    const { attributeNames, children, aggregates } = this.parseFields();
+    const totalCount = this.data.result.records.length;
+    this.soqlQuery(attributeNames, this.massageRows(this.data.result.records, children, aggregates), totalCount);
+  }
+
+  public parseFields(): ParsedFields {
+    const fields = this.columns;
+    // Field names
+    const attributeNames: string[] = [];
+
+    // For subqueries. Display the children under the parents
+    const children: string[] = [];
+
+    // For function fields, like avg(total).
+    const aggregates: FunctionField[] = [];
+
+    if (fields) {
+      this.logger.info(`Found fields ${JSON.stringify(fields.map((field) => `${typeof field}.${field.name}`))}`);
+
+      fields.forEach((field) => {
+        if (field instanceof SubqueryField) {
+          children.push(field.name);
+          field.fields
+            .map((subfield: Field) => subfield as SubqueryField)
+            .forEach((subfield: SubqueryField) => attributeNames.push(`${field.name}.${subfield.name}`));
+        } else if (field instanceof FunctionField) {
+          if (field.alias) {
+            attributeNames.push(field.alias);
+          } else {
+            attributeNames.push(field.name);
+          }
+          aggregates.push(field);
+        } else {
+          attributeNames.push(field.name);
+        }
+      });
+    } else {
+      this.logger.info(`No fields found for query "${this.data.query}"`);
+    }
+
+    return { attributeNames, children, aggregates };
+  }
+
+  /* eslint-disable @typescript-eslint/no-explicit-any */
+  public soqlQuery(columns: string[], records: any[], totalCount: number): void {
```

Comment:
can we create a `record` type? it seems it's passed as any[] into most functions - just nice to add types where we can

Reply:
@WillieRuemmele, I will take a look at it.